File: 0101-0200/0169.Majority_Element.cpp
//0ms






class Solution {
public:
    int majorityElement(vector<int>& nums) {
        map<int,int> map;
        for(int i=0;i<nums.size();i++){
            map[nums[i]]++;
        }
        for(auto it:map){
            if(it.second > nums.size()/2){
                return it.first;
            }
        }
        return 0;
    }
};






//0ms





class Solution { //Moore's Voting Algorithm
public:
    int majorityElement(vector<int>& nums) {
         // Size of the given array
        int n = nums.size();
        // Count
        int cnt = 0;
        // Element
        int el; 
        // Applying the algorithm
        for (int i = 0; i < n; i++) {
            if (cnt == 0) {
                cnt = 1;
                el = nums[i];
            } else if (el == nums[i]) {
                cnt++;
            } else {
                cnt--;
            }
        }
        /* Checking if the stored element
         is the majority element*/
        int cnt1 = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] == el) {
                cnt1++;
            }
        }
        //return element if it is a majority element
        if (cnt1 > (n / 2)) {
            return el;
        }
        //return -1 if no such element found
        return -1;
    }
};

File: 0101-0200/0169.Majority_Element.rs
//0ms





impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        nums.into_iter()
            .fold((0, -1), |(count, mut candidate), num| {
                if count == 0 {
                    candidate = num;
                }
                (count + if candidate == num { 1 } else { -1 }, candidate)
            })
            .1
    }
}





//0ms






impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut candidate = nums[0];
        let mut count = 1;
        for i in 1..nums.len() {
            if count == 0 {
                candidate = nums[i];
                count += 1;
            }
            else if candidate == nums[i] {
                count += 1;
            } else {
                count -= 1;
            }
        }
        candidate
    }
}
